RFC: Add copy_file_range support to bsdcpio

luis-henrix opened the pull request:
This is an RFC so that I get some feed on adding support for this (linux-only!) syscall.  I'll need to add checks in configure.ac (and to cmake files) so that this code gets compiled out for other platforms.  But for now I just wanted to make sure this is something maintainers would be interested in.

luis-henrix commented:
I've just pushed another version that should (hopefully!) pass all the tests.  The only change is the addition of a test to check if copy_file_range exists on the system.

luis-henrix commented:
Any feedback?  Anyone?  I'm just trying to understand if this is something worth spending more time on.  Maybe I should have added more details on my initial comment.  Here's the commit message:

> copy_file_range syscall allow the underlying filesystems to provide
> acceleration techniques such as the use of reflinks.  This may improve
> performance and optimize space.

kientzle commented on cpio/cpio.c:
```diff
@@ -865,24 +904,8 @@ entry_to_archive(struct cpio *cpio, struct archive_entry *entry)
-		}
-	}
+	if (r >= ARCHIVE_WARN && archive_entry_size(entry) > 0 && fd >= 0)
+		copy_file(cpio, fd);
```
If we can line up the API right, this should just be a call to `archive_write_disk_copy_file`.  We should always prefer to have a clean public API and use that as directly as possible.

Looking at this more closely makes me wonder about the right structure here:
* The easiest API to explain would be one where you first call `archive_write_header` to write the header and then `archive_write_contents_from_file(archive, srcpath)` to write the body of that entry.
* But as the comments above explain, that complicates error handling; if the source file cannot be opened, we would like to be able to just skip this entry entirely (not writing the header).  That's why the original code was written to copy contents from a file descriptor.
* Exposing `copy_file()` from a file descriptor raises a lot of questions about the full meaning of that call (for example, what if the file descriptor is not at the beginning of the file?).

So I'm a little torn:  copying from a file descriptor enables the right kind of error handling, but copying from a filename is a cleaner API.  Thoughts?

luis-henrix replied:
> Looking at this more closely makes me wonder about the right structure here:
> 
>    * The easiest API to explain would be one where you first call `archive_write_header` to write the header and then `archive_write_contents_from_file(archive, srcpath)` to write the body of that entry.

Agreed!

>    * But as the comments above explain, that complicates error handling; if the source file cannot be opened, we would like to be able to just skip this entry entirely (not writing the header).  That's why the original code was written to copy contents from a file descriptor.

Yeah, I re-wrote my code a bunch of times to try to handle every error path.  The API isn't perfect, I agree.  I guess it could use the srcpath instead of a file descriptor but, apart from the extra pain in error handling opening the file, the original file descriptor would be also required.  I.e. we would have to open that file twice: once for the header and the second time in the public API.

>    * Exposing `copy_file()` from a file descriptor raises a lot of questions about the full meaning of that call (for example, what if the file descriptor is not at the beginning of the file?).

Right, my implementation doesn't really care about that as it simply assumes the FDs (both of them) are in the right position.  The API will attempt to copy X bytes from the current fd_in to the current fd_out. 

> So I'm a little torn: copying from a file descriptor enables the right kind of error handling, but copying from a filename is a cleaner API. Thoughts?

I would rather not re-open the file in the API as this would mean yet another syscall.  But I understand your point.

kientzle commented on libarchive/archive.h:
```diff
@@ -944,6 +944,9 @@ __LA_DECL int archive_write_disk_set_user_lookup(struct archive *,
 __LA_DECL la_int64_t archive_write_disk_gid(struct archive *, const char *, la_int64_t);
 __LA_DECL la_int64_t archive_write_disk_uid(struct archive *, const char *, la_int64_t);
+#ifdef HAVE_COPY_FILE_RANGE
+__LA_DECL la_ssize_t archive_write_disk_copy_file(struct archive *, int, off_t);
```
I like this new method, but if we're going to add it, it should make sense as a reusable service (not just as a cpio helper).
* What if the `fd_in` is not currently positioned at the beginning of the file?  What should happen?
* Should this function take an fd and size or should it take a filename instead?
* Clearly, we can provide it regardless of whether the platform supports this particular sys call.  (If other platforms provide similar services, we should plan to support them in the future as well.)
* The name could be more accurate:  It does not "copy a file"; it fills the current entry with the contents of the specified file.

luis-henrix replied:
> I like this new method, but if we're going to add it, it should make sense as a reusable service (not just as a cpio helper).
> 
>   * What if the `fd_in` is not currently positioned at the beginning of the file?  What should happen?

As I said above, the function doesn't really care about that, it simply copies X bytes from one file to the other.  I guess that, if the final version of the API will still receive a file descriptor, this would need to be very well documented.

>   * Should this function take an fd and size or should it take a filename instead?

Again, as I said above, this would mean an extra syscall as cpio would already need that file open, right?  Should it be closed before?

>   * Clearly, we can provide it regardless of whether the platform supports this particular sys call.  (If other platforms provide similar services, we should plan to support them in the future as well.)

Right, moving the copy loop into the API as you suggest below would be a solution: the caller wouldn't need to bother knowing whether copy_file_range exists or not. 

>   * The name could be more accurate:  It does not "copy a file"; it fills the current entry with contents from the specified file.

Ok, I'll try to thing of something else.  Naming is hard :-)

kientzle commented on libarchive/archive_write_disk_posix.c:
```diff
@@ -2429,6 +2429,70 @@ create_filesystem_object(struct archive_write_disk *a)
 	return (0);
 }
 
+#ifdef HAVE_COPY_FILE_RANGE
```
This also requires `SEEK_DATA` support.

luis-henrix replied:
Ups!  I'll add that.

kientzle commented on cpio/cpio.c:
```diff
@@ -782,13 +782,52 @@ file_to_archive(struct cpio *cpio, const char *srcpath)
+		lafe_errc(1, archive_errno(cpio->archive), "%s",
+			  archive_error_string(cpio->archive));
+#endif
+	/* Fallback to a copy loop in case of partial copy or if copy failed */
```
This next block of code should be part of `archive_write_disk_copy_file`.

luis-henrix replied:
Yeah, to be honest that was my initial intention.  I tried to do that initially but found out that things would be a bit more complicated.  For example, the loop uses the cpio->buff and ->buff_size; moving it into `archive_write_disk_copy_file` would mean to re-implement that copy logic.  Obviously, not a show-stopper.  I'll rethink it and try to come up with something different.

kientzle commented:
This looks pretty intriguing and would be a nice speedup.  Your proposal to add a new public API to fill a new file with data from an existing file is appealing, but I'm not yet convinced that the interface you have here is quite right.  (I'm not certain it's _wrong_, either.  I just need to think about it a bit more.)

luis-henrix commented:
I'll think a bit more about it and try to propose something different.  The first thing I'll need to do is to have a closer look at the whole API (which I'm not really familiar with, I'm afraid).  This will allow me to get a better feeling on how to define a new one.  Then, I'll need to understand what are the consequences of using a filename for the API.  The call from the cpio code actually needs that file description before **and** after the copy, which makes things a bit more tricky.

Anyway, thanks a lot for your review @kientzle, very good input from you.  And I'll see if I'm able to have v2 of this patch during next week.

kientzle commented:
After giving it some thought, I think the API that takes a file descriptor and size is probably the best approach.  So let's stick with that design.

luis-henrix commented:
Hi @kientzle I've done some rework on this patch and my current version has a slightly different interface (and a bit more ugly).  Here's the function signature:

` la_ssize_t
archive_write_disk_copy_file(struct archive *_a, void *buff, size_t buff_size,
                            int fd_in, off_t s)
`

These new parameters are required for the copy loop, which will be moved out from cpio.  I guess the alternative would be to have a new buffer inside the library.  In the call from cpio the `buff` and `buff_size` are taken respectively from `cpio->buff` and `cpio->buff_size`.  Thoughts?

Another question: would you prefer to have me pushing new (incremental) commits on top of what's on my `copy-file-range` branch, or are you OK with a `git push --force`?

luis-henrix commented:
Ok, I went ahead and pushed a new version of the patch.  I've also added a new test for checking sparse files copy (this test is probably is a bit too hacky, but I can work on it a bit more).

luis-henrix commented:
As I mentioned above, an alternative to the API is to keep it as

`
la_ssize_t archive_write_disk_copy_file(struct archive *_a, int fd_in, off_t s)
`

and have the function allocate (in the stack) the buffer.  This would allow to drop the `buff` and `buff_size` fields in `struct cpio`.  I can push such a version really quick if you prefer that @kientzle

luis-henrix commented:
And I just pushed yet another revision:
1. dropped `buff` and `buff_size` from `struct cpio` and use local buffers in `archive_write_disk_copy_file`
2. also added API to `archive_write_disk_windows.c`, which should fix the windows build failures (I hope, but can't verify locally)

luis-henrix commented:
@kientzle ping? :-)

luis-henrix commented:
Hey @kientzle, sorry for insisting, but I was wondering if I'm missing something else for this feature.  Did you had a chance of looking at the new version I pushed last month?

patmantru commented:
<bump> ?

emaste commented:
> support for this (linux-only!) syscall

FreeBSD also has copy_file_range (now).
```
HISTORY
     The copy_file_range() function appeared in FreeBSD 13.0.
```
